okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -170,32 +172,44 @@ private RealConnection findConnection(int connectTimeout, int readTimeout, int w
         refusedStreamCount = 0;
       }
     }
-    RealConnection newConnection = new RealConnection(connectionPool, selectedRoute);
+
+    RealConnection result = new RealConnection(connectionPool, selectedRoute);
 
     synchronized (connectionPool) {
-      acquire(newConnection);
-      Internal.instance.put(connectionPool, newConnection);
-      this.connection = newConnection;
+      acquire(result);
+      Internal.instance.put(connectionPool, result);
```

Comment:
Oooh, yeah that’ll break things. So we allow a stream allocation to hold a connection before it’s in the pool? That’s not too terrible. I think this code needs a pass of simplification – it’s too subtle.